Mongo - Postgres migration - core side

jackyalbo opened the pull request:
Signed-off-by: jackyalbo <jacky.albo@gmail.com>

### Explain the changes
1. 

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

jackyalbo committed:
core side - db migration

Signed-off-by: jackyalbo <jacky.albo@gmail.com>

dannyzaken reviewed (commented):
1. maybe separate the general migration script to another file. it will be easier to test and reuse. you can then just pass any collections list to the generic migration script regardless of which DBs are involved.

2. for the source DB (mongo) don't pass the DB indexes - you don't want to cause it to start building indexes right before the migration.

4. we should test if it's worth it to migrate several collections in parallel. 
5. before starting to migrate a collection drop the collection in the `to_client` to get rid of failed attempts.
6. store in the DB the status for each collection, so you can skip them on retries.
7. for the large collections (chunks, blocks, etc.) we should consider storing retry markers so we won't need to restart the entire collection on failures.
8. in `migrate_collection` catch errors and retry a few times before failing the entire migration
9. check if the query need to be sorted

jackyalbo commented:
@dannyzaken I addressed some of your requests and added a basic test. Still WIP but want your feedback (and to check test is passing). Thanks.

dannyzaken commented on src/test/unit_tests/test_postgres_upgrade.js:
```diff
@@ -0,0 +1,250 @@
+    ],
+    properties: {
+        _id: { objectid: true },
+        key1: { type: 'string' },
```
maybe use more types in the test other than string

jackyalbo replied:
will do

dannyzaken commented on src/test/unit_tests/test_postgres_upgrade.js:
```diff
@@ -0,0 +1,250 @@
+    async find(query, options = {}) {
+        const start = (query && query._id && query._id.$gt) || 0;
+        const end = options.limit ? start + options.limit : this.size;
+        console.log("data to be moved", this.data.slice(start, end), query);
```
won't it make the test very chatty? I think it's better to print only on failures

jackyalbo replied:
was for debugging - removed

dannyzaken commented on src/upgrade/migrator.js:
```diff
@@ -0,0 +1,88 @@
+        this.upgrade_table = this.to_client.collection('migrate_status');
+        this.migrate_status = await this.upgrade_table.findOne({});
+        if (!this.migrate_status) {
+            this.migrate_status = {
```
add a schema to the `migrate_status` collection. postgres_client uses the schema to encode\decode the json

jackyalbo replied:
done

dannyzaken commented on src/upgrade/migrator.js:
```diff
@@ -0,0 +1,88 @@
+        if (!this.migrate_status) {
+            this.migrate_status = {
+                _id: this.to_client.new_object_id(),
+                collection_index: 0,
```
add more info to help with debugging - collection_name, write_date, total_size, num_inserted, etc.

jackyalbo replied:
done

dannyzaken commented on src/upgrade/migrator.js:
```diff
@@ -0,0 +1,88 @@
+                done = true;
+            }
+        }
+        console.log(`completed migration of ${total} documents to table ${col_name}`);
```
as a sanity check, maybe check that the sizes of the collections are the same

jackyalbo replied:
done

liranmauda commented:
/retest ci/centos/unit

jackyalbo committed:
fixes to pg_client

schema for decoding
check for duplicate idx
stats to parse as ints instead of strings

Signed-off-by: jackyalbo <jacky.albo@gmail.com>
